bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/i18n/KNXTranslationProvider.java

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright (c) 2010-2022 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.knx.internal.i18n;
+
+import java.text.MessageFormat;
+import java.util.Locale;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.i18n.LocaleProvider;
+import org.openhab.core.i18n.TranslationProvider;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.FrameworkUtil;
+
+/**
+ * This class provides translations. It is a helper class for i18n / localization efforts.
+ *
+ * @implNote It is implemented as a static singelton, enforced by the single-element enum pattern.
+ * @apiNote @set() must be called to provide tanslation service, otherwise all functions will return untranslated text.
+ *          Thread safety is ensured.
+ * @author Holger Friedrich - Initial contribution
+ *
+ */
+@NonNullByDefault
+public enum KNXTranslationProvider {
```

Comment:
Why is this implemented as **enum**? You could consider implementing it as a **class** and inject `LocaleProvider` and `TranslationProvider` in constructor. This way you would also avoid some nullness issues. See for example:
https://github.com/openhab/openhab-addons/blob/6e4f78458d143ced35485241ab1ec87c0990dcc9/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/MieleTranslationProvider.java#L29-L60

Reply:
I wanted to implement the TranslationProvider as a static singleton. This makes it easy to use from any class within the binding. Injecting LocaleProvider and TranslationProvider is done in KNXHandlerFactory, which hands it over to the static TranslationProvider instance.

I tried implement it differently before by propagating LocaleProvider and TranslationProvider all the way down from KNXHandlerFactory to all classes which require the translation service.... This was quite a mess and the static singleton approach looks much clearer.

>  * @implNote It is implemented as a static singelton, enforced by the single-element enum pattern.

Implementing a singleton using an enum is actually a quite nice pattern from the boot "Effective Java". It may look a bit strange when reading the code the first time :-)

Reply:
OK.